include/verilated_vpi.cpp

```diff
@@ -646,23 +645,22 @@ class VerilatedVpiImp final {
                 continue;
             }
             VerilatedVpiCbHolder& ho = *it++;
-            if (VerilatedVpioVar* const varop = VerilatedVpioVar::castp(ho.cb_datap()->obj)) {
-                void* const newDatap = varop->varDatap();
-                void* const prevDatap
-                    = varop->prevDatap();  // Was malloced when we added the callback
-                VL_DEBUG_IF_PLI(VL_DBG_MSGF("- vpi: value_test %s v[0]=%d/%d %p %p\n",
-                                            varop->fullname(), *(static_cast<CData*>(newDatap)),
-                                            *(static_cast<CData*>(prevDatap)), newDatap,
-                                            prevDatap););
-                if (std::memcmp(prevDatap, newDatap, varop->entSize()) != 0) {
-                    VL_DEBUG_IF_PLI(VL_DBG_MSGF("- vpi: value_callback %" PRId64 " %s v[0]=%d\n",
-                                                ho.id(), varop->fullname(),
-                                                *(static_cast<CData*>(newDatap))););
-                    update.insert(varop);
-                    vpi_get_value(ho.cb_datap()->obj, ho.cb_datap()->value);
-                    (ho.cb_rtnp())(ho.cb_datap());
-                    called = true;
-                }
+            VerilatedVpioVar* const varop = reinterpret_cast<VerilatedVpioVar*>(ho.cb_datap()->obj);
```

Comment:
I agree with your point we can speed this up by removing the dynamic cast, but we probably should have an assertion before this, in case the assumption changes or something is broken.  E.g.

      // Value callbacks should only registered against VpioVars, but assert that...
      VL_DEBUG_IF_PLI(assert(dynamic_cast<VerilatedVpioVar*>(ho.cb_datap()->obj)););

Perhaps we should also assert where adding to the cbObjList.

Reply:
Done!

> Perhaps we should also assert where adding to the cbObjList.

I will need to check for the kind of `reason` and assert only for `cbValueChange`. Would you like to put something like this before [this](https://github.com/verilator/verilator/blob/3e986517c314ce68938b60c52273fd29f47b24bf/include/verilated_vpi.cpp#L1396) like?
```
  // Value callbacks should only registered against VpioVars, but assert that...
  VL_DEBUG_IF_PLI(assert(reason == cbValueChange && dynamic_cast<VerilatedVpioVar*>(vop->castVpiHandle())););
  ```

Reply:
I was going to add it, but noticed that we actually already have this check when we add the callback:
https://github.com/verilator/verilator/blob/3e986517c314ce68938b60c52273fd29f47b24bf/include/verilated_vpi.cpp#L532